utils/prof_utils.py

```diff
@@ -0,0 +1,523 @@
+
+    Parameters:
+    -----------
+    profile: list
+        The profile to find the widths of
+    
+    Return: 
+    W10: float
+        The W10 width of the profile measured in number of bins
+    W50: float
+        The W50 width of the profile measured in number of bins
+    Weq: float
+        The equivalent width of the profile measured in number of bins   
+    """
+    #perform spline operations
+    x = np.array(list(range(len(profile))))
+    spline0 = UnivariateSpline(x, profile, s=0)
+    spline10 = UnivariateSpline(x, profile - np.full(len(x), 0.1), s=0)
+    spline50 = UnivariateSpline(x, profile - np.full(len(x), 0.5), s=0)
+
+    #find Weq
+    integral = spline0.integral(0, len(profile)-1)
+    Weq = integral/max(profile)
+
+    #find W10 and W50
+    W10_roots = spline10.roots()
+    W50_roots = spline50.roots()
+    W10=0
+    W50=0
+    for i in range(0, len(W10_roots), 2):
+        W10 += W10_roots[i+1] - W10_roots[i]
+    for i in range(0, len(W50_roots), 2):
+        W50 += W50_roots[i+1] - W50_roots[i]
+
+    return W10, W50, Weq
+
+def regularize_prof(profile, reg_param=5e-8):
+    """
+    Applies the following operations to a pulse profile:
+    1) Sigma Clip
+    2) Find the on-pulse bins from Sigma Clip
+    3) Fill unnecessary nans from the clipping process
+    4) Apply Stickel 2010 regularization
+    5) Normalize the profile
+    6) Set the off-pulse to zero
+
+    Parameters:
+    -----------
+    profile: list
+        The pulse profile
+    reg_param: float
+        OPTIONAL - the regularization parameter used for smoothing the profile. Default:5e-8
+    
+    Returns:
+    --------
+    on_pulse_prof: list
+        The pulse profile after the above operations
+    noise_mean: float
+        The mean of the profile noise
+    """
+    #clip the profile to find the on-pulse
+    noise_std, clipped_prof = sigmaClip(profile)
+    #Reverse the nan-profile. Then replace nans with zero:
+    on_pulse_prof = []
+    on_pulse_bins = []
+    for i, val in enumerate(clipped_prof):
+        if np.isnan(val):
+            on_pulse_prof.append(profile[i])
+            on_pulse_bins.append(i)
+        else:
+            on_pulse_prof.append(0)
+
+    #Some noisy profiles have nans in the middle of the actual pulse... fix:
+    on_pulse_prof = fill_clipped_prof(on_pulse_prof, profile)
+
+    #regularization with Stickel - smoothing
+    x = np.linspace(0,len(profile)-1, len(profile), dtype=int)
+    y = np.array(on_pulse_prof)
+    data = np.column_stack((x, y))
+    stkl = Stickel(data)
+    stkl.smooth_y(reg_param)
+    on_pulse_prof = stkl.yhat
+
+    #subract noise mean and normalize profile
+    noise_mean = np.nanmean(clipped_prof)
+    on_pulse_prof = np.array(on_pulse_prof) - noise_mean
+    on_pulse_prof = on_pulse_prof/np.max(on_pulse_prof)
+    #reset noise to 0
+    for i, val in enumerate(on_pulse_prof):
+        if val<0:
+            on_pulse_prof[i]=0
+        
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Trailing whitespace](https://app.codacy.com/app/CIRA-Pulsars-and-Transients-Group/vcstools/pullRequest?prid=4645369)